storage/repository/repository.go

```diff
@@ -0,0 +1,262 @@
+// storage.Copy() is called, the storage package will internally look up
+// the repository associated with the scheme of the URI, then it will use
+// a type assertion to check if the repository implements CopyableRepository.
+// If so, the Copy() function will be run from the repository, otherwise
+// storage.Copy() will return NotSupportedError. This works similarly for
+// all other methods in repository-related interfaces.
+//
+// Note that a repository can be registered for multiple URI schemes. In such
+// cases, the repository must internally select and implement the correct
+// behavior for each URI scheme.
+//
+// A repository will only ever need to handle URIs with schemes for which it
+// was registered, with the exception that functions with more than 1 operand
+// such as Copy() and Move(), in which cases only the first operand is
+// guaranteed to match a scheme for which the repository is registered.
+//
+// NOTE: most developers who use Fyne should *not* generally attempt to
+// call repository methods directly. You should use the methods in the storage
+// package, which will automatically detect the scheme of a URI and call into
+// the appropriate repository.
+//
+// Since: 2.0.0
+type Repository interface {
+
+	// Exists will be used to implement calls to storage.Exists() for the
+	// registered scheme of this repository.
+	//
+	// Since: 2.0.0
+	Exists(u fyne.URI) (bool, error)
+
+	// Reader will be used to implement calls to storage.Reader()
+	// for the registered scheme of this repository.
+	//
+	// Since: 2.0.0
+	Reader(u fyne.URI) (fyne.URIReadCloser, error)
+
+	// CanRead will be used to implement calls to storage.CanRead() for the
+	// registered scheme of this repository.
+	//
+	// Since: 2.0.0
+	CanRead(u fyne.URI) (bool, error)
+
+	// Destroy is called when the repository is un-registered from a given
+	// URI scheme.
+	//
+	// The string parameter will be the URI scheme that the repository was
+	// registered for. This may be useful for repositories that need to
+	// handle more than one URI scheme internally.
+	//
+	// Since: 2.0.0
+	Destroy(string)
+}
+
+// CustomURIRepository is an extension of the repository interface which
+// allows the behavior of storage.ParseURI to be overridden. This is only
+// needed if you wish to generate custom URI types, rather than using Fyne's
+// URI implementation and net/url based parsing.
+//
+// NOTE: even for URIs with non-RFC3986-compliant encoding, the URI MUST begin
+// with 'scheme:', or storage.ParseURI() will not be able to determine which
+// storage repository to delegate to for parsing.
+//
+// Since: 2.0.0
+type CustomURIRepository interface {
+	Repository
+
+	// ParseURI will be used to implement calls to storage.ParseURI()
+	// for the registered scheme of this repository.
+	ParseURI(string) (fyne.URI, error)
+}
+
+// WriteableRepository is an extension of the Repository interface which also
```

Comment:
typo: s/Writeable/Writable

This incorrect spelling has been used several times throughout this PR and the wiki spec, but I don't think it would be helpful for me to comment on every one so I'll just add this one but please apply everywhere.

Reply:
cf845be9c5af83786bc20f6f823654e00b3c1ced